### DISCUSSION TITLE
Github Enterprise integration

### DISCUSSION THREAD

**Discussion by atse** [zed-industries/zed#8278] [Feature Requests]
Local collaboration is currently tied to the public instance of GitHub [^1].

- Has anyone tried directing the calls to their on-prem instance?
- Any plans to make this configurable and supported (as a commercial offering, I'd imagine)?

[^1] https://zed.dev/docs/local-collaboration

**Comment by corrydiehl** [comment]
I"m interested in this as well as my company uses an Enterprise instance of Github in the cloud and this would make leadership much more amenable to collaboration if it was tied to our own instance.

**Comment by kimroen** [comment]
We would also really like this! In vscode we use Copilot with github enterprise, but in Zed I haven't found a way of enabling this.

**Reply by 1925381584** [reply]
everything is coming after git and debugger

**Reply by kimroen** [reply]
Git is released now, so only debugger to go I suppose? 😄

**Comment by AspadaX** [comment]
I could set this up in earlier versions but after I signed out for once, I cannot do this anymore. Does anyone know how to do that.

**Reply by kimroen** [reply]
So you're saying this worked in a previous version? If so, sounds like maybe we should post this as a bug/regression instead of a discussion.

**Reply by AspadaX** [reply]
> So you're saying this worked in a previous version? If so, sounds like maybe we should post this as a bug/regression instead of a discussion.

I think we can. But I had upgraded Zed and I forgot which version it was. Maybe I should install an older version and see if it works with GHE.

**Comment by vegardx** [comment]
GitHub Enterprise Cloud (also with Data Residency) just needs you to set the Base URL (eg. foobar.ghe.com, where foobar is the GitHub Enterprise). Zed just need to expose this in the settings, as they use the same API schema for GitHub.com and GHE.com. The only odd-ball would be GitHub Enterprise Server, which uses a different set of APIs, but I don't think they've launched things like GitHub Copilot for GitHub Enterprise Server anyway.

I hope this can be solved soon, as I'd love to use Zed as my default editor at work.

**Reply by AspadaX** [reply]
> GitHub Enterprise Cloud (also with Data Residency) just needs you to set the Base URL (eg. foobar.ghe.com, where foobar is the GitHub Enterprise). Zed just need to expose this in the settings, as they use the same API schema for GitHub.com and GHE.com. The only odd-ball would be GitHub Enterprise Server, which uses a different set of APIs, but I don't think they've launched things like GitHub Copilot for GitHub Enterprise Server anyway.
> 
> I hope this can be solved soon, as I'd love to use Zed as my default editor at work.

Yes. I think so. Would like to see a PR that addresses this issue.

**Comment by AspadaX** [comment]
[Link to potentially relevant source code](https://github.com/zed-industries/zed/blob/d0f7dede799140082fa2e56ed891cc1e603fbacc/crates/copilot/src/copilot_chat.rs#L205)

If I inferred correctly, I think exposing the the base url out and then instantiate a client from the `new` method of `CopilotChat` could address the issue? 

```rust
impl CopilotChat {
    pub fn global(cx: &App) -> Option<gpui::Entity<Self>> {
        cx.try_global::<GlobalCopilotChat>()
            .map(|model| model.0.clone())
    }

    pub fn new(fs: Arc<dyn Fs>, client: Arc<dyn HttpClient>, cx: &App) -> Self {
        let config_paths = copilot_chat_config_paths();

        let resolve_config_path = {
            let fs = fs.clone();
            async move {
                for config_path in config_paths.iter() {
                    if fs.metadata(config_path).await.is_ok_and(|v| v.is_some()) {
                        return config_path.clone();
                    }
                }
                config_paths[0].clone()
            }
        };

        cx.spawn(|cx| async move {
            let config_file = resolve_config_path.await;
            let mut config_file_rx = watch_config_file(cx.background_executor(), fs, config_file);

            while let Some(contents) = config_file_rx.next().await {
                let oauth_token = extract_oauth_token(contents);

                cx.update(|cx| {
                    if let Some(this) = Self::global(cx).as_ref() {
                        this.update(cx, |this, cx| {
                            this.oauth_token = oauth_token;
                            cx.notify();
                        });
                    }
                })?;
            }
            anyhow::Ok(())
        })
        .detach_and_log_err(cx);

        Self {
            oauth_token: None,
            api_token: None,
            client,
        }
    }
```

**Comment by lironsam** [comment]
Any update on this ?

**Reply by AspadaX** [reply]
> Any update on this ?

I found the solution a couple of days ago, you may refer to the follow doc:
https://zed.dev/docs/ai/edit-prediction#using-github-copilot-enterprise

**Reply by Jdahn** [reply]
@AspadaX, does this approach still work for you? For me the login flow always starts at github.com even when applying the settings mentioned in your link.

**Reply by Andreas02-dev** [reply]
> @AspadaX, does this approach still work for you? For me the login flow always starts at github.com even when applying the settings mentioned in your link.

```json
  "edit_predictions": {
    "provider": "copilot",
    "copilot": { "enterprise_uri": "https://foobar.ghe.com" },
  },
```

This works for me on Zed 1.3.6 (Version: 1.3.6+stable).

**Reply by Jdahn** [reply]
@Andreas02-dev  I tried your configuration. but the login flow still points me to github.com instead of "https://XXX.ghe.com"
i am using version 1.9.0+stable.316.ced90fc636c4ede05402befc38a63bae7fd741bd

**Reply by ruudschouten** [reply]
Worked for me, but after changing the settings you have to login via `Settings > AI > LLM Providers > GitHub Copilot Chat > that button there`
It should then use your set enterprise url.

**Reply by michael-meindl** [reply]
I can get edit predictions to work with Enterprise, but normal agent chat is still not possible. Settings > AI > LLM Providers > GitHub Copilot Chat > login button still tries to access github.com.
Any way to configure that in the settings file?
